cmd/utils/hierarchical_coordinator.go

```diff
@@ -405,53 +606,277 @@ func (hc *HierarchicalCoordinator) ChainEventLoop(chainEvent chan core.ChainEven
-				lru.Add(head.Block.Hash(), node)
-				hc.recentBlocks[head.Block.Location().Name()] = lru
-				hc.recentBlockMu.Unlock()
+		case <-sub.Err():
+			return
+		}
+	}
+}
+
+func (hc *HierarchicalCoordinator) MapConstructProc() {
+	defer func() {
+		if r := recover(); r != nil {
+			log.Global.WithFields(log.Fields{
+				"error":      r,
+				"stacktrace": string(debug.Stack()),
+			}).Fatal("Go-Quai Panicked")
+		}
+	}()
+	defer hc.wg.Done()
+
+	for {
+		select {
+		case <-hc.pendingHeaderBackupCh:
+			log.Global.Error("Running the backup calculation")
+			hc.PendingHeadersMap()
+		case <-hc.quitCh:
+			return
+		default:
+		}
+	}
+}
+
+func (hc *HierarchicalCoordinator) ComputeMapPending(head core.ChainEvent) {
+	defer func() {
+		if r := recover(); r != nil {
+			log.Global.WithFields(log.Fields{
+				"error":      r,
+				"stacktrace": string(debug.Stack()),
+			}).Fatal("Go-Quai Panicked")
+		}
+	}()
+	backend := hc.GetBackend(head.Block.Location())
+	entropy := backend.TotalLogEntropy(head.Block)
+	node := Node{
+		hash:     head.Block.Hash(),
+		number:   head.Block.NumberArray(),
+		entropy:  entropy,
+		location: head.Block.Location(),
+	}
+	hc.recentBlockMu.Lock()
+	defer hc.recentBlockMu.Unlock()
+	locationCache, exists := hc.recentBlocks[head.Block.Location().Name()]
+	if !exists {
+		// create a new lru and add this block
+		lru, _ := lru.New[common.Hash, Node](c_recentBlockCacheSize)
+		lru.Add(head.Block.Hash(), node)
+		hc.recentBlocks[head.Block.Location().Name()] = lru
+		log.Global.WithFields(log.Fields{"Hash": head.Block.Hash(), "Number": head.Block.NumberArray()}).Debug("Received a chain event and calling build pending headers")
+	} else {
+		bestBlockHash := locationCache.Keys()[len(locationCache.Keys())-1]
+		_, exists := locationCache.Peek(bestBlockHash)
+		if exists {
+			oldestBlockHash := locationCache.Keys()[0]
+			oldestBlock, exists := locationCache.Peek(oldestBlockHash)
+			if exists && oldestBlock.entropy.Cmp(node.entropy) < 0 {
+				locationCache.Add(head.Block.Hash(), node)
+				hc.recentBlocks[head.Block.Location().Name()] = locationCache
 				log.Global.WithFields(log.Fields{"Hash": head.Block.Hash(), "Number": head.Block.NumberArray()}).Debug("Received a chain event and calling build pending headers")
-				hc.BuildPendingHeaders(stopChan)
+			}
+		}
+	}
+}
+
+func (hc *HierarchicalCoordinator) PendingHeadersMap() {
+
+	hc.recentBlockMu.Lock()
+	defer hc.recentBlockMu.Unlock()
+	var badHashes map[common.Hash]bool
+	badHashes = make(map[common.Hash]bool)
+	count := 0
+	var leaders []Node
+search:
+	if count > 0 {
+		circularShift(leaders)
+		badHashes = make(map[common.Hash]bool)
+	}
+
+	if count > 2 {
+		log.Global.Error("Too many iterations in the build pending headers, skipping generate")
+		return
+	}
+	// Pick the leader among all the slices
+	backend := *hc.consensus.GetBackend(common.Location{0, 0})
+	defaultGenesisHash := backend.Config().DefaultGenesisHash
```

Comment:
Why set default to cryprus1?

Reply:
Will make an issue for when we do hierarchical expansions.